distributed/worker_memory.py

```diff
@@ -255,7 +257,13 @@ async def _maybe_spill(self, worker: Worker, memory: int) -> None:
 
             total_spilled += weight
             count += 1
-            await asyncio.sleep(0)
+            # If the current buffer is filled with a lot of small values,
+            # evicting one at a time is very slow and the worker might
+            # generate new data faster than it is able to evict. Therefore,
+            # only pass on control if we spent at least 0.5s evicting
+            if monotonic() - start > 0.5:
+                await asyncio.sleep(0)
+                start = monotonic()
```

Comment:
I'm confused by both the old and the new solution here.  It seems like we're staying stuck in a while loop burning the CPU at 100%.  Is this correct?  If so, I'm against this design generally.  (although I'm not objecting to this change, since it doesn't necessarily make things worse from my perspective.  

Is my understanding correct here that we're blocking the event loop?

Reply:
My understanding was flawed here.  @crusaderky and @fjetter set me straight here.  I didn't realize that `data.evict` was blocking.

Reply:
Yes, it's blocking the event loop. No, it's not burning CPU at 100%.

I think the idea here is that it's a crude form of asyncio scheduling. You're basically saying "stop everything! there's nothing that could possibly be more important to do (launch tasks, fetch data, send data, talk to scheduler, etc.) than dumping data to disk right now." That's a bit of a dramatic approach, but not too wrong for this situation. Spilling to disk should probably be the highest priority thing we do at this moment. And ideally starting new tasks, fetching new data, or anything else that could produce memory probably should pause until eviction is done.

This only works because we don't have async disk https://github.com/dask/distributed/issues/4424. With async writes, we'd have to figure out how to actually pause other subsystems without blocking the whole event loop in this situation.

Reply:
Yeah, I don't disagree with any of that.  My previous understanding was flawed.  Please ignore my previous comment.  I'm good here.

I'm totally fine passing control periodically.

Reply:
> No, it's not burning CPU at 100%.

This is hardly surprising - when a write() is stuck waiting on a slow network file system, it shouldn't burn CPU.